---
title: "Copy number variation across the WatSeq dataset"
author: Ricardo H. Ramirez-Gonzalez
date: 23 Sept 2020
output: 
    revealjs::revealjs_presentation:
    self_contained: true 
---
<style type="text/css" media="screen">

.left {
  float: left;
  width:65%;
  z-index:-10;
}

.right {
  
  float: right;
  justify-content: center;
  display: table-cell;
  vertical-align: middle;
  padding-top:70px;
  z-index:-10;
  width:35%;
}

</style>

# Detection of medium and large CNV

Objective

- To find medium and large Copy Number Variation (CNV)

```{r, message=FALSE, echo=FALSE, warning=FALSE}

params<-list()
params$data_path <- "~/Documents/WatSeq/Deletions/20200724/200bp/"
params$gff_file <- "~/Dropbox/JIC/WatSeq/Annotation/Triticum_aestivum.IWGSC.41.parts.tidy.gff3.gz"
params$gene_in_detail <- "TraesCS4D02G040100"


library(devtools)
library(knitr)
#library(GenomicRanges)
library(GenomicFeatures)
library(RSQLite)
library(ggbio)
library(plyr)
library(dplyr) 
library(rtracklayer)
load_all("../bio.cnv/R", TRUE)
gff  <- import(params$gff_file)

seq_len=paste0(params$data_path, "sequence_lengths.txt")
#levels <- c("chr1A","chr1B", "chr1D","chr2A","chr2B", "chr2D","chr3A","chr3B", "chr3D","chr4A","chr4B", "chr4D","chr5A","chr5B", "chr5D","chr6A","chr6B", "chr6D","chr7A","chr7B", "chr7D","chrUn")
lens <- read.csv(seq_len, sep="\t", stringsAsFactors=F)
levels <- lens$seqname
lens$seqname<-factor(lens$seqname, levels = levels)

covs_db = dbConnect(drv=RSQLite::SQLite(), dbname="~/Documents//WatSeq/Deletions/20200724/200bp/long_tables/covs_200bp.db")
gene_region <- gff[gff$gene_id %in% c(params$gene_in_detail)]

cov_in_range <- getWindowsInRange(covs_db, region=gene_region)

txdb <- makeTxDbFromGFF(file="~/Dropbox/JIC/WatSeq/Annotation/Triticum_aestivum.IWGSC.41.parts.tidy.gff3")
```


# Initial data 

![](images/01.CNV-pipeline_bg.png){ width=100% }

```{r, message=FALSE, echo=FALSE}
mat <- read.csv("./examples/mock_for_presentation.csv",)
rownames(mat) <- mat[,1]
mat[,1] <- NULL
data_df <- getExonsDF(mat)
kable(mat)

```

## Initial data 

```{r, message=FALSE, echo=FALSE, fig.width=10, fig.height=5}

p <- plotLibraryOnChromosomeInterval(cov_in_range, lines=c(), col="cov", title = "Original coverage" ) + coord_cartesian(ylim=c(0,60)) 
p <- p + theme(legend.position="bottom")
p
```

> - Coverage across regions a region is not uniform

# Normalisation

![](images/02.CNV-pipeline_bg.png){ width=100% }

$x_{i,j}=\frac{WindowCoverage_{i,j}\times10^{9}}{WindowLength_{i}\times{totalReadsSample_{j}}}$

$xnorm_{i,j}=\frac{x_{i,j}}{mean(X_{i})}$


## Normalise by sample

$x_{i,j}=\frac{WindowCoverage_{i,j}\times10^{9}}{WindowLength_{i}\times{totalReadsSample_{j}}}$

. . . 

```{r, message=FALSE, echo=FALSE}


totalReadsPerSample<-apply(mat,2,sum, na.rm=T)
tmp_mat <- rbind(mat, totalReadsPerSample) 
rownames(tmp_mat)[6] <- "totalReadsSample"
kable(tmp_mat)
```

## Normalisation by window 

$xnorm_{i,j}=\frac{x_{i,j}}{mean(X_{i})}$

. . . 

```{r, message=FALSE, echo=FALSE}
multiplier<-outer(data_df$ExonL, totalReadsPerSample )
multiplier<-1/multiplier
multiplier<-1000000000*multiplier
#"RPKM-like covs"
multiplier<-mat*multiplier
mean.mult <- apply(multiplier, 1, mean, na.rm = T)
tmp_mat <- cbind(multiplier, mean.mult ) 
colnames(tmp_mat)[5] <- "Window Mean"
kable(tmp_mat)
#kable(multiplier)
```

## Normalised coverage

```{r, message=FALSE, echo=FALSE}
covs<-sweep(multiplier,MARGIN=1,mean.mult,'/')
kable(covs)
```

## Normalised coverage

```{r, message=FALSE, echo=FALSE, fig.width=11, fig.height=3}
p <- plotLibraryOnChromosomeInterval(cov_in_range, lines=c(), col="cov", title = "Original coverage" ) + coord_cartesian(ylim=c(0,60)) 
p <- p + theme(legend.position="bottom")
p
```

. . . 

```{r, message=FALSE, echo=FALSE, fig.width=11, fig.height=3}
p <- plotLibraryOnChromosomeInterval(cov_in_range, lines=c(), col="norm_cov", title = "Normalised coverage" ) + coord_cartesian(ylim=c(0,2)) 
p <- p + theme(legend.position="bottom")
p
```


## Second normalisation

> - Remove the windows with 
$sd(window) > 0.3$
> - Repeat normalisation  
> - Exclude the datapoints with "0"

## Standard deviation of lines across samples.  

::: {.left }
```{r, message=FALSE, echo=FALSE, warning=FALSE,  fig.align="center"}
sd_path <- paste0(params$data_path, "libSD_short.csv")
libsd <- read.csv(sd_path)
noisy_lines <- libsd[libsd$SD > 0.45,]
plotHistogram(libsd, column = "SD",     binwidth=0.01, trim_range=F)
```
:::

. . .

::: {.right}

``r nrow(noisy_lines)`` out of ``r nrow(libsd) `` lines are noisy ($\sigma > 0.45$). 
```{r, message=FALSE, echo=FALSE, warning=FALSE, fig.align="center"}
noisy_lines %>% mutate_if(is.numeric, format, digits=2) %>% kable( . , row.names=FALSE)
```
:::


## Regular vs noisy line 


```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
p <- plotLibraryOnChromosomeInterval(cov_in_range, lines=c("WATDE0214","WATDE0039"), col="norm_cov", title = "Normalised coverage" ) + coord_cartesian(ylim=c(0,2)) 
p <- p + theme(legend.position="bottom")
p
```

## Regular line 

```{r, message=FALSE, echo=FALSE,  warning=FALSE,fig.width=11, fig.height=6}
lines <- c("WATDE0214")
gene_region_flank <- flank(gene_region, 1500000, both=TRUE)
plotRegionScatterLines(covs_db, gff, gene_region_flank, lines=lines, show.ribons=T, show.cov=T, genes=c(params$gene_in_detail), shape=".")

```

## Noisy line 

```{r, message=FALSE, echo=FALSE, warning=FALSE, fig.width=11, fig.height=6}
lines <- c("WATDE0039")
plotRegionScatterLines(covs_db, gff, gene_region_flank, lines=lines, show.ribons=T, show.cov=T, genes=c(params$gene_in_detail), shape=".")

```

# Merging continous lines 

```{r, message=FALSE, echo=FALSE,  warning=FALSE}

cnvs_to_plot <- read.csv(file = "./examples/sample_regions.csv")
colnames(cnvs_to_plot)[1] <- "seqnames"
#kable(cnvs_to_plot)
cnvs_to_plot <- makeGRangesFromDataFrame(cnvs_to_plot, keep.extra.columns = T)

cnvs_to_plot_flanking <- flank(cnvs_to_plot, 300000, both=TRUE)

#cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M10/test_all_cnv.csv")
#cnvs_df <- read.csv(cnvs_file)

cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M10/all_cnv.csv.gz") 
cnvs_df <- read.csv(gzfile(cnvs_file))
cnvs <- makeGRangesFromDataFrame(cnvs_df, keep.extra.columns=T)
cnvs<-fixRangesOrder(cnvs, lens, ordr=lens$seqname)
```


- Individual CNVs events may be informative, but we want to find large events


## Stich CNV candidates 

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
plotRegionScatterLines(covs_db, gff, cnvs_to_plot[1], lines=cnvs_to_plot[1]$line, show.ribons=T, 
show.cov=T, shape=".", stiched.cnv = cnvs , txdb = txdb )

```


## Stich CNV candidates 

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
plotRegionScatterLines(covs_db, gff, cnvs_to_plot[2], lines=cnvs_to_plot[2]$line, show.ribons=T, 
show.cov=T, shape=".", stiched.cnv = cnvs , txdb = txdb )

#print("hi")
```


## Stich CNV candidates 

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
plotRegionScatterLines(covs_db, gff, cnvs_to_plot[3], lines=cnvs_to_plot[3]$line, show.ribons=T, 
show.cov=T, shape=".", stiched.cnv = cnvs, txdb = txdb )

#print("hi")
```


## Stich CNV candidates 

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
plotRegionScatterLines(covs_db, gff, cnvs_to_plot[4], lines=cnvs_to_plot[4]$line, show.ribons=T, 
show.cov=T, shape=".", stiched.cnv = cnvs, txdb = txdb )

#print("hi")
```


## Stich CNV candidates 

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
p.to.extend <- plotRegionScatterLines(covs_db, gff, cnvs_to_plot[5], lines=cnvs_to_plot[5]$line, show.ribons=T, 
show.cov=T, shape=".", stiched.cnv = cnvs )
p.to.extend
#print("hi")
```


# CNVs across the full genome 

![](images/03.CNV-pipeline_bg.png){ width=100% }

Can we get an overview across a line? 



```{r, message=FALSE, echo=FALSE,  warning=FALSE}
large_cnvs <- cnvs_df[cnvs_df$width > 1500,]

total_dels <- nrow(cnvs_df)
over_1500 <- nrow(large_cnvs)


non_singletons_cnvs <- cnvs_df[cnvs_df$width > 200,]

non_singletons_cnvs_gr <-  makeGRangesFromDataFrame(cnvs_df, keep.extra.columns=T)
non_singletons_cnvs_gr<-fixRangesOrder(non_singletons_cnvs_gr, lens, ordr=lens$seqname)

over_200 <- nrow(non_singletons_cnvs)
percentage_over_200 <- 100.0 * over_200 / total_dels

percentage_over_1500 <- 100.0 * over_1500 / total_dels
lines_in_file <- unique(cnvs_df$line)
```


## CNV length distributions

There are ```r format(total_dels, nsmall=0, big.mark=",")``` CNV events across   ```r length(lines_in_file)``` lines


```{r, echo=FALSE,  warning=FALSE, message=FALSE}
plotHistogram(cnvs_df, column="width", binwidth=200,trim_range = F) + xlim(c(0,5000))
```


## CNV length distributions (over 200bp)

> - The minimum size that we have in this dataset is 200bp ```r format(over_200, nsmall=0, big.mark="," )``` (```r format(percentage_over_200, digits=4, big.mark="," )``` %) are not singletons 

. . .

```{r, echo=FALSE,  warning=FALSE, message=FALSE}
plotHistogram(non_singletons_cnvs, column="width", binwidth=200)

```

## Deletion in 4B may be incomplete

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
p.to.extend
#print("hi")
```

## Explore global view

```{r, echo=FALSE,  warning=FALSE, message=FALSE}
plotCnvsInLine(non_singletons_cnvs_gr, line = "WATDE0214")
```

## Deletion in full 

```{r, message=FALSE, echo=FALSE,  warning=FALSE, fig.width=11, fig.height=6}
plotRegionScatterLines(covs_db, gff, cnvs_to_plot[6], lines=cnvs_to_plot[6]$line, show.ribons=T, 
show.cov=T, shape=".", stiched.cnv = cnvs )


```

# Next steps

> - Improve stiching algorithm 
> - Use smaller window size (150bp, 100bp)
> - Find genes/regions more prone to have CNVs
> - Analyse in detail known CNVs 





