---
output: 
    pdf_document

params: 
    set_title: "Global analysis of CNV"
    data_path: "~/Documents/WatSeq/Deletions/20200724/200bp/"
    gff: "~/Dropbox/JIC/WatSeq/Annotation/Triticum_aestivum.IWGSC.41.parts.tidy.gff3.gz"
title: "`r params$set_title`"
author: "Ricardo H. Ramirez-Gonzalez"
---

```{r, message=FALSE, echo=FALSE}

# params<-list()
# params$data_path <- "~/Documents/WatSeq/Deletions/20200724/200bp/"
# params$gff <- "~/Dropbox/JIC/WatSeq/Annotation/Triticum_aestivum.IWGSC.41.parts.tidy.gff3.gz"
#params$gff <- "~/Dropbox/JIC/WatSeq/Annotation/IWGSCv1.0_UTR_ALL.parts.tidy.gff3.gz"

library(devtools)
library(knitr)
library(GenomicRanges)
library(GenomicFeatures)
library(RSQLite)
library(ggbio)
library(rtracklayer)
require(plyr)
require(dplyr)
library("parallel")
load_all("../bio.cnv/R", TRUE)
gff  <- import(params$gff)

seq_len=paste0(params$data_path, "sequence_lengths.txt")
#levels <- c("chr1A","chr1B", "chr1D","chr2A","chr2B", "chr2D","chr3A","chr3B", "chr3D","chr4A","chr4B", "chr4D","chr5A","chr5B", "chr5D","chr6A","chr6B", "chr6D","chr7A","chr7B", "chr7D","chrUn")
lens <- read.csv(seq_len, sep="\t", stringsAsFactors=F)
levels <- lens$seqname
lens$seqname<-factor(lens$seqname, levels = levels)

db_path <- paste0(params$data_path, "/long_tables/covs_200bp.db")
covs_db = dbConnect(drv=RSQLite::SQLite(), dbname=db_path)


#cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M10/all_cnv.csv.gz") 
#cnvs_df <- read.csv(gzfile(cnvs_file))

#cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M10/all_cnv.csv") 
#cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M10/chr6D_part2.cnv.for.db.tsv") 
cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M10/chr6D_part2.cnv.for.db.tsv") 
cnvs_df_M10 <- read.csv(cnvs_file, sep = "\t" )

cnvs_file <- paste0(params$data_path, "/long_tables/cnv_out_M3/chr6D_part2.cnv.for.db.tsv") 
cnvs_df_M3 <- read.csv(cnvs_file, sep = "\t" )

cnvs_M3 <- makeGRangesFromDataFrame(cnvs_df_M3, keep.extra.columns=T)
cnvs_M10 <- makeGRangesFromDataFrame(cnvs_df_M10, keep.extra.columns=T)

#without_name <- cnvs[is.na(seqnames(cnvs))]
cnvs_M3<-fixRangesOrder(cnvs_M3, lens, ordr=lens$seqname)
cnvs_M10<-fixRangesOrder(cnvs_M10, lens, ordr=lens$seqname)

genes_to_display <- c("TraesCS6D02G366600", "TraesCS6D02G400000", "TraesCS6D02G375500", "TraesCS6D02G395100", "TraesCS6D02G401400")

to_ignore <- c("WATDE0009", "WATDE0039",  "WATDE0056", "WATDE0060" ,  "WATDE0090")

cnvs_M10 <- cnvs_M10[cnvs_M10$line %in% setdiff(unique(cnvs_M10$line), to_ignore) ]
cnvs_M3 <- cnvs_M3[cnvs_M3$line %in% setdiff(unique(cnvs_M3$line), to_ignore) ]

filled_M3  <- fillAllCNVsAroundGenes(cnvs_M3, gff[gff$type=="gene"], mc.cores = 4)

#filled <- filled[filled$line %in% setdiff(unique(filled$line), to_ignore) ]

# gene_to_show <- "TraesCS1B02G049300"
# region_1B <- gff[gff$gene_id %in% c(gene_to_show)] 
# region_1B <- flank(region_1B, 5000, both=TRUE)


sample_lines <- c(
    "WATDE0663", "WATDE0721", "WATDE0444", "WATDE0057", "WATDE0687", "WATDE0632", "WATDE1023", 
    "WATDE0544", "WATDE0079", "WATDE1046", "WATDE0759", "WATDE1009", "WATDE0944", "WATDE0038", "WATDE1017", "WATDE0926",
    "WATDE0059", "WATDE0606", "WATDE0543", "WATDE0503", "WATDE1049", "WATDE0638", "WATDE0730", "WATDE0941", "WATDE1032", 
    "WATDE1016", "WATDE0817", "WATDE0109", "WATDE0176", "WATDE0691", "WATDE0456", "WATDE0453", "WATDE1018", "WATDE0561", 
    "WATDE0704", "WATDE0945", "WATDE0781", "WATDE0975", "WATDE0910")

genes_to_display_regions <- gff[gff$gene_id %in% genes_to_display]
genes_to_display_regions <- exetendGR(genes_to_display_regions, size=2000)

first_gene <- genes_to_display_regions[1]
#filled_lines_values <- subsetByOverlaps(cnvs, first_gene)
#lines_with_changes <- unique(filled_lines_values[filled_lines_values$cnv_level != 1]$line)
#sample_lines <-  sample(lines_with_changes, 20)
region_copy_M3 <- countCNVsPerRegion(cnvs_df_M3)
region_copy_M10 <- countCNVsPerRegion(cnvs_df_M10)
````

 
First we want to explore some lines. 



```{r,message=FALSE, echo=FALSE, fig.width=7, fig.height=7,  warning=FALSE, results='asis'}
p1 <- plotHistogram(region_copy_M10, column="n", title_prefix="Max Gap: 10 (wrong) ",trim_range = T)
p2 <- plotHistogram(region_copy_M10[region_copy_M10$n>1,], column="n", title_prefix="Max Gap: 10 (wrong) ",trim_range = F,binwidth = 1) + coord_cartesian(xlim=c(0,50))
p3 <- plotHistogram(region_copy_M3, column="n", title_prefix="Max Gap: 3 ",trim_range = T)
p4 <- plotHistogram(region_copy_M3[region_copy_M3$n>1,], column="n", title_prefix="Max Gap: 3 ",trim_range = F,binwidth = 1) + coord_cartesian(xlim=c(0,50))
multiplot(plotlist= list(p1,p2,p3,p4), cols = 2)
```


```{r,  message=FALSE, echo=FALSE, fig.width=7, fig.height=7,  warning=FALSE, results='asis'}
#plotDeletionsInRegion(cnvs, genes_to_display_regions[2], lines=sample_lines)
lines_per_plot <- 4
regions_test <- read.csv("6d_regions_to_test.csv")
regions_test <- makeGRangesFromDataFrame(regions_test)

for(i in seq(1, length(regions_test))){
    print(plotDeletionsInRegion(filled_M3, regions_test[i], lines=sample_lines))
    for(j in seq(1,  length(sample_lines)/lines_per_plot ) ){

        first_line <- j * lines_per_plot
        last_line  <- first_line + lines_per_plot
        last_line <- ifelse(last_line > length(sample_lines), length(sample_lines), last_line)
        lines_j <- sample_lines[seq(first_line, last_line)]  
        p1 <- plotRegionScatterLines(covs_db, gff, 
            regions_test[i], 
            lines=lines_j, 
            stiched.cnv = filled_M3, 
            show.ribons=T, 
            norm_cov_range=c(0,4 ))
        print(p1)
    }

}

```



